### ISSUE TITLE
Backup and restore by SQL+file copy results in file content ENODATA

### ISSUE THREAD

**Issue by CrowhopTech** [supabase/storage-api#314]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

To implement a backup/restore strategy for my local Supabase install, I created a backup tool which stops all Supabase services, copies out the contents of `/var/lib/storage` and all contents of the Postgres database, and then starts them all up again.
Restore does the same but inverse (shutdown, apply SQL dump and copy storage files in).

After using this process, everything appears in Studio, Postgres, and the filesystem correctly, but requests to get the actual data return in an `ENODATA` error (see below). Newly inserted files function just fine and I can get the content.

> ```
> storage {"level":"info","time":"2023-04-18T05:57:56.153Z","pid":19,"hostname":"shinysorter-dev-supabase-storage-cbbc47c97-875pv","reqId":"req-6","tenantId":"stub","project":"stub","msg":"stub/files/Rain landing at ArcCorp!.mp4"}
> storage {"level":"info","time":"2023-04-18T05:57:56.200Z","pid":19,"hostname":"shinysorter-dev-supabase-storage-cbbc47c97-875pv","reqId":"req-6","tenantId":"stub","project":"stub","req":{"method":"GET","url":"/object/public/files/Rain%20landing%20at%20ArcCorp!.mp4","headers":{"host":"supabase> storage:5000","x_forwarded_proto":"http","x_real_ip":"10.42.0.1","user_agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36","accept":"*/*","referer":"http://192.168.1.5/"},"hostname":"supabase-storage:5000","remoteAddress":"10.42.0.115","remotePort":37398},"res":{"statusCode":500},"responseTime":59.58428382873535,"error":{"code":"ENODATA","errno":61},"msg":"GET | 500 | 10.42.0.115 | req-6 | /object/public/files/Rain%20landing%20at%20ArcCorp!.mp4 | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36"}
> ```

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Upload a file to Supabase storage
2. Stop all Supabase services, dump Postgres to a file, and copy contents of Supabase storage
3. Delete the Supabase install to simulate a full site loss
4. Reinstall Supabase and run the restore
5. Note that the files are present in the Supabase Studio UI and the database/filesystem entries are correct, but attempting to view the file results in the above error in the Storage logs and a 500 error for the client

## Expected behavior

The files should be usable just like they were pre-restore, and API calls should succeed to get the data.

## System information

- OS: [e.g. macOS, Windows]: Windows, Supabase services running on Ubuntu
- Browser (if applies) [e.g. chrome, safari]: Chrome
- Version of supabase-js: [e.g. 6.0.2]: 2.2.1
- Version of Node.js: [e.g. 10.10.0]: 18.7.1

## Additional context

I'm running all of these services in a Kubernetes cluster (self-developed chart, not a public one): will happily disclose any more details required about the setup. Many other features (including most storage operations without backup/restore) have been working just fine so I got it mostly right.

**Comment by fenos** [comment]
Hello, @CrowhopTech the tool you created for backups seems amazing!

This is interesting... Are you sure you are attaching the correct volume to the container when restoring the files?

**Comment by pckimlong** [comment]
I have the same issue

**Comment by CrowhopTech** [comment]
Oh sorry! I somehow missed that first comment @fenos. Yes, I had the same thought, so I ran `ls` in `/var/lib/storage` and found the usual tree there (`/var/lib/storage/stub/stub/<bucket name>/<file name>`)

**Comment by fenos** [comment]
@CrowhopTech what happens if you try to keep the database as is and delete the folder bucket, then re-apply a copy of the files in the same location?

**Comment by CrowhopTech** [comment]
Ooooh good idea! Will do

**Comment by CrowhopTech** [comment]
I think this debugging exercise just helped me figure it out: `tar` doesn't preserve extended attributes by default, you need to add the `--xattrs` flag. I remember that being a big deal while I was setting up my Supabase Storage install, so I'm not surprised that causes issues here. Let me test that...

Testing steps:
1. Added file `Star Citizen 1_9_2022 8_52_35 AM` to Supabase Storage using the Studio UI
2. Validated that the file is present in the storage container:
```
/app # ls -l /var/lib/storage/stub/stub/files/Star\ Citizen\ 1_9_2022\ 8_52_35\ AM.png
-rw-r--r--    1 root     root       8106457 Apr 21 07:42 /var/lib/storage/stub/stub/files/Star Citizen 1_9_2022 8_52_35 AM.png
```
3. Validated the entry in the database:
```
"id","bucket_id","name","owner","created_at","updated_at","last_accessed_at","metadata","path_tokens"
"0123a7a7-c05e-458b-969b-d4f582223e8f","files","Star Citizen 1_9_2022 8_52_35 AM.png","","2023-04-21 07:42:05.579013+00","2023-04-21 07:42:05.765855+00","2023-04-21 07:42:05.579013+00","[object Object]","Star Citizen 1_9_2022 8_52_35 AM.png"
```
4. Took a backup using the same tool for simplicity
5. In the filesystem, deleted the directory: `rm -rf /var/lib/storage/stub`. Validated that studio no longer works to render the file. Of note is that the logs now return a different error, `ENOENT`, instead of `ENODATA`:
```
storage {"level":"info","time":"2023-04-21T07:47:30.861Z","pid":18,"hostname":"shinysorter-dev-supabase-storage-cbbc47c97-rmg9b","reqId":"req-6","tenantId":"stub","project":"stub","msg":"stub/files/Star Citizen 1_9_2022 8_52_35 AM.png"}                                                      
storage {"level":"info","time":"2023-04-21T07:47:30.863Z","pid":18,"hostname":"shinysorter-dev-supabase-storage-cbbc47c97-rmg9b","reqId":"req-6","tenantId":"stub","project":"stub","req":{"method":"GET","url":"/object/public/files/Star%20Citizen%201_9_2022%208_52_35%20AM.png?t=2023-04-21T07:47:29.638Z","headers":{"host":"supabase-storage:5000","x_forwarded_proto":"http","x_real_ip":"10.42.0.1","user_agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36","accept":"image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8","referer":"http://192.168.1.5:32690/"},"hostname":"supabase-storage:5000","remoteAddress":"10.42.0.132","remotePort":58560},"res":{"statusCode":500},"responseTime":13.028529644012451,"error":{"errno":-2,"code":"ENOENT","syscall":"open","path":"/var/lib/storage/stub/stub/files/Star Citizen 1_9_2022 8_52_35 AM.png"},"msg":"GET | 500 | 10.42.0.132 | req-6 | /object/public/files/Star%20Citizen%201_9_2022%208_52_35%20AM.png?t=2023-04-21T07:47:29.638Z | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36"}
```
6. Re-created the directory (`mkdir -p /var/lib/storage/stub/stub/files`) and added the file back from my backup
```
/app # ls /var/lib/storage/stub/stub/files -l
total 7920
-rw-r--r--    1 root     root       8106457 Apr 21 07:51 Star Citizen 1_9_2022 8_52_35 AM.png
```
7. After adding the file back, we return to the original error of `ENODATA` instead of `ENOENT`.

**Comment by CrowhopTech** [comment]
That was it! Storing the file attributes using `getfattr -R <storage directory> --absolute-names > storage-attrs` and then restoring using `setfattr --restore=storage-attrs` worked! My guess is the `ENODATA` error means "no data in the attributes", and hence why it was a different error from the `ENOENT`.

For those curious, these are the attributes on the file.
```
# file: supabase-storage//stub/stub/files/Star Citizen 1_9_2022 9_46_58 AM.png
user.supabase.cache-control=""
user.supabase.content-type=""
```

Closing this as I'm able to successfully restore and view files after restoration. Thanks for the pointer!

**Comment by CrowhopTech** [comment]
@pckimlong see the response above, and let me know if you have any questions! I seem to have successfully gotten it to work, so I'd love to help you out too

**Comment by CrowhopTech** [comment]
Also for those curious, here is my code (mainly `backup.ts` and `restore.ts`). It's somewhat specialized to my install but the general idea should be usable for others: it just makes a tarball of a SQL dump, the storage directory, and the xattrs of the storage directory. Restore simply does the inverse.

**Comment by pckimlong** [comment]
> @pckimlong see the response above, and let me know if you have any questions! I seem to have successfully gotten it to work, so I'd love to help you out too

In my situation, I wanted to move my database to another computer, but I couldn't use pg_dump because the old computer couldn't access the internet and didn't have pg_dump installed. So, I copied the entire Supabase directory, including the volume mounted to supabase-storage, to another computer. When I started the Supabase instance, all my data in the database was lost because I hadn't copied the database volume yet. So, I stopped the Supabase containers, copied the database volume from my old computer, replaced it in the new computer's volume directory, and then started Supabase again. This time, all the data persisted. However, like you, I could see the files represented in the Supabase studio UI but was unable to view or download them. When I tried to download them, I received an error status code 500 with the message 'internal server error'. When I tested by uploading another file, I was able to view and download it normally.
